jmclient/jmclient/wallet_service.py

```diff
@@ -200,9 +202,12 @@ def register_callbacks(self, callbacks, txinfo, cb_type="all"):
             # note that in this case, txid is ignored.
             self.callbacks["all"].extend(callbacks)
         elif cb_type in ["unconfirmed", "confirmed"]:
-            if txinfo not in self.callbacks[cb_type]:
-                self.callbacks[cb_type][txinfo] = []
-            self.callbacks[cb_type][txinfo].extend(callbacks)
+            if callbacks:
+                reg = self.callbacks[cb_type].setdefault(txinfo, [])
+                if isinstance(reg, str):
+                    # found a txid breadcrumb for this txinfo
+                    reg = self.callbacks[cb_type].setdefault(reg, [])
+                reg.extend(callbacks)
```

Comment:
So I'm really struggling with this edit for multiple reasons:

1. the initial `if callbacks:` (line 205) was surprising, we never currently call this function with an empty list, I can't see any reason to do so?

2. The two uses of `setdefault` were hard for me to parse. Initially just because I'd never seen `setdefault` before, but, once I thought I'd understood it, I still failed to parse this code. So I understand that for a dict `x`, `x.setdefault(a, b)` will set the value for key `a` to `b` if `a` is not currently present, and otherwise will leave the value for `a` untouched, and it will return the value for key `a` (whether updated, or not).

But here I get lost: on line 206 you take that return value `reg`, which must be a list (no?), i.e. specifically the list of callbacks for that key `txinfo`, and then you check if it is `isinstance` `str`?

I guess I must have got something quite badly wrong.

And a final more general point:

3. If this is not changing function, only syntax, why do it, since it is longer? Do you see a performance reason? I can't imagine that'd be the case since we're only parsing very small lists/dicts here?

Just in case it helps the discussion, here's how I envision the structure of these callback dicts:

```
self.callbacks = {"unconfirmed": {"sometxid1": [callbackfn1, callbackfn2],
                                                    (sometxoutputtuple): [callbackfn3, callbackfn4],
                                                     "sometxid2": []}
                          "confirmed": {"sometxid3": [callbackfn5],
                                                (sometxoutputtuple): [callbackfn6, callbackfn7]}
                          "all": [callbackfn8, callbackfn9]}
```

Reply:
The `if callbacks` is just to avoid adding an empty list as a value in the callbacks dictionary since then there'd be nothing that would ever cause it to be cleaned out.

The `if isinstance` is to redirect the registration of a callback under a provisional key so that the callback is instead registered under the corresponding actual TxID key.

Reply:
Old:
```
self.callbacks = {"unconfirmed": {"sometxid1": [callbackfn1, callbackfn2],
                                  (sometxoutputtuple): [callbackfn3, callbackfn4],
                                  "sometxid2": []},
                  "confirmed":   {"sometxid3": [callbackfn5],
                                  (sometxoutputtuple): [callbackfn6, callbackfn7]},
                  "all":         [callbackfn8, callbackfn9]}
```

New:
```
self.callbacks = {"unconfirmed": {"sometxid1": [callbackfn1, callbackfn2],
                                  (unseentxoutputtuple2): [callbackfn3, callbackfn4],
                                  (seentxoutputtuple1): "sometxid1"},
                  "confirmed":   {"sometxid3": [callbackfn5],
                                  (unseentxoutputtuple3): [callbackfn6, callbackfn7]},
                  "all":         [callbackfn8, callbackfn9]}
```
* The value of a TxID key is always a list of callback functions.
* The value of a tx-outputs tuple key may be either a list of callback functions (if the corresponding transaction has not yet been observed in the Bitcoin Core wallet) or a string specifying the TxID of the observed transaction.
* There should never be any empty callback lists in `self.callbacks` (aside from an empty `all` list), as these are essentially the garbage-collected equivalent of memory leaks.